compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -1222,6 +1222,76 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
         Ok(())
     }
+
+    pub fn swap_memory(
+        &mut self,
+        x_ptr: Pointer<Option<M::Provenance>>,
+        y_ptr: Pointer<Option<M::Provenance>>,
+        count: u64,
+        layout: ty::layout::TyAndLayout<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let size = layout.size;
+        let align = layout.align.abi;
+        let Some((x_alloc_id, x_offset, _)) = self.get_ptr_access(x_ptr, size, align)? else {
+            // Called on ZST so it is noop.
+            return Ok(())
+        };
+        let Some((y_alloc_id, y_offset, _)) = self.get_ptr_access(y_ptr, size, align)? else {
+            unreachable!("If right param is ZST, left must be too")
+        };
+
+        if count > i64::MAX as u64 {
+            throw_ub_format!("`count` argument to `swap_nonoverlapping_many` is too large.");
+        }
+
+        if x_alloc_id == y_alloc_id {
+            if (x_offset..x_offset + size * count).contains(&y_offset)
+                || (y_offset..y_offset + size * count).contains(&x_offset)
+            {
+                throw_ub_format!("swap was called on overlapping memory.");
+            }
+        }
+        if count == 0 {
+            return Ok(());
+        }
+
+        let tmp_stack_alloc = {
+            let can_use_values = layout.ty.is_primitive() || layout.ty.is_slice();
+            if can_use_values {
+                None
+            } else {
+                let tmp_x_place = self.allocate(layout, MemoryKind::Stack)?;
+                let tmp_y_place = self.allocate(layout, MemoryKind::Stack)?;
+                Some((tmp_x_place, tmp_y_place))
+            }
+        };
+
+        for i in 0..i64::try_from(count).unwrap() {
+            let curr_x_ptr = self.ptr_offset_inbounds(x_ptr, layout.ty, i)?;
+            let curr_y_ptr = self.ptr_offset_inbounds(y_ptr, layout.ty, i)?;
+
+            let x_mplace = MPlaceTy::from_aligned_ptr(curr_x_ptr, layout);
+            let y_mplace = MPlaceTy::from_aligned_ptr(curr_y_ptr, layout);
+            if let Some((tmp_x_place, tmp_y_place)) = tmp_stack_alloc {
+                self.copy_op(&x_mplace.into(), &tmp_x_place.into(), false)?;
```

Comment:
This is doing a *typed* swap! Is that really what we want? If it is, then this is the wrong file. `memory.rs` is for byte-level, untyped operations. This should probably be a function in `intrinsics.rs` then. Also it should take places as arguments, not pointers, if that is the level of abstraction it acts upon.

Reply:
Judging from [these docs](https://doc.rust-lang.org/nightly/std/ptr/fn.swap_nonoverlapping.html) the copy probably should be untyped. That means `memory.rs` is the right file and taking pointers makes sense. But it means you must not use `copy_op` in here as that would have too much UB. Use `mem_copy` instead.